### ISSUE TITLE
GLUE Finetuning script

### ISSUE THREAD

**Issue by chenmoneygithub** [keras-team/keras-nlp#400]
There are several questions about the GLUE finetuning script:
1. GLUE finetuning is basically backbone model + classification head. Shall we make this classification head a standalone API, maybe named as `GlueClassifier`? 
2. If YES to 1, can we assume the `GlueClassifier` takes in two args about backbone model? 1) backbone model instance 2) a string identifier tells how to extract the classifier input from the backbone model's output, e.g., "pooled_output" for BERT.
3. For the script itself, do we want to have distribution strategy support, like multi-gpu or TPU? 
4. Do we want to let users be able to run the script with their model in a single command? This is tough because we need to pipe the backbone model instance to the script. 

My thoughts on these questions:
1. I would like to have `GlueClassifier` instead of each backbone model using its own classification head for GLUE.
2. yes.
3. Yes.
4. I would reply on users to copy the code and run, and we can have explicit message on which part should be customized to reflect the custom model.

Here is a draft colab: https://colab.research.google.com/gist/chenmoneygithub/9c6dfd911759ecdf8edf110054619378/glue-finetuning.ipynb

**Comment by jbischof** [comment]
It appears `GlueClassifier` is essentially the same as `BertClassifier`. Why couldn't you reuse `BertClassifier`?

**Comment by chenmoneygithub** [comment]
@jbischof This is my question 1) (should've been more clear) - do we want a reusable `GlueClassifier` or let users create their own classifier for each model?

**Comment by jbischof** [comment]
Forgive me for not following but if you replaced your custom classifier with a `BertClassifier` I'm pretty sure it would run OOB! Creating a glue script is helpful but I don't see why we would need extra modeling code to accomplish this.

**Comment by mattdangerw** [comment]
Yeah, I think trying to write single classification head for all models will not generalize well. Your current implementation in the colab involves indexing into a dictionary output with to a "pooled" dense representation of some sort.

 - This will not work for RoBERTa, which has no pooled output by default. In fact, we don't even output a dictionary! The classifier for RoBERTa applies [two dense layers](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/models/roberta/roberta_tasks.py#L70-L80) to index 0 of the sequence embedding.
 - This would also not work if you wanted to simple average pool over an entire sequence.
 - For things like t5, an output string label is auto-regressively sampled from a language model output, and then converted back to an integer label.

We might also want to do some thinking about the different ways this could be run. The BERT paper describes a learning rate hyperparameter search, so there is no single learning rate for the real BERT GLUE score. And for large models these tasks can take many hours, so you might want to distribute runs across many workers.

I think all of this is to say maybe we should start with a script, get it working across a good cross section of models, and then take a look if there is any "library code" to pull out of it.

**Comment by chenmoneygithub** [comment]
@jbischof It's about reusability. This GLUE finetuning script is expected to be able to easily adapt to custom models. If we don't provide a classification head, then users need to do. 

@mattdangerw Yea, relying on a string identifier is not quite stable. I am kinda unsure about how this classification part should be provided to our users. I want it to be inline with the script because it gives a clear clue what the finetuning part does, so users can adapt to their models. 

Agree we should start simple. I am thinking about using BERT, RoBERTa and GPT to create the story, and we can evaluate from there where we can improve.

**Comment by jbischof** [comment]
@chenmoneygithub following the design pattern of the library so far, we should expect each model to provide their own classification layer because, as @mattdangerw detailed nicely, each model may have a unique implementation. We will provide implementations for all models in our library, e.g., `BertClassifier`, `RobertaClassifier` etc. 

I think the input to this script should be a classification layer and not a raw encoder. Tell the user which args are expected and they can write a wrapper if necessary.

**Comment by chenmoneygithub** [comment]
Just got back! Resuming this discussion:

@jbischof I am hoping users can use this script on their custom models. It looks trivial that they just need to implement a simple classification head, but without explicitly told so, it could take users (especially our basic users) a while to figure out this fact, that's why I want to have the classification head explicitly coded up in the script.

@jbischof @mattdangerw I think it is safe to assume that the users' model could always output a vector representation for the input sequence, so my proposal is to explicitly document that the `ClassificationHead` should take the vector representation as input. For the shipped code, I will just showcase with our pretrained BERT, WDYT?

**Comment by chenmoneygithub** [comment]
To reflect our latest discussion - we will delete the classification head code from the script.

**Comment by chenmoneygithub** [comment]
Forgive my back and forth please, but putting in more thoughts i still find the usefulness of having a explicit ClassificationHead users can use. So this is my new plan: use a flag `use_default_classification_head`, setting which to `True` causes the eval to use the default classification head, otherwise it is using the custom head such like `BertClassifier`. I am going to make a PR and we can discuss from there.
